chalk-solve/src/clauses.rs

```diff
@@ -104,6 +104,69 @@ pub fn push_auto_trait_impls<I: Interner>(
     });
 }
 
+/// Leak auto traits for opaque types, just like `push_auto_trait_impls` does for structs.
+///
+/// For example, given the following program:
+///
+/// ```notrust
+/// #[auto] trait Send { }
+/// trait Trait { }
+/// struct Bar { }
+/// opaque type Foo: Trait = Bar
+/// ```
+/// Checking the goal `Foo: Send` would generate the following:
+///
+/// ```notrust
+/// Foo: Send :- Bar: Send
+/// ```
+pub fn push_auto_trait_impls_opaque<I: Interner>(
+    builder: &mut ClauseBuilder<'_, I>,
+    auto_trait_id: TraitId<I>,
+    opaque_id: OpaqueTyId<I>,
+) {
+    debug_heading!(
+        "push_auto_trait_impls_opaque({:?}, {:?})",
+        auto_trait_id,
+        opaque_id
+    );
+
+    let opaque_ty_datum = &builder.db.opaque_ty_data(opaque_id);
```

Comment:
Reminder that we should factor the "hidden type" out from this struct and make it something that we lazilly request

Reply:
How would that work, actually? A method on the database to request it, or something similar?

Reply:
Yeah, just a separate method on the db that only gets invoked when auto trait impls are actually required (that's important) would be fine, I think